### PR TITLE
Qt fixes for RHEL6/7

### LINKED ISSUE #4960: Unable to build qt on CentOS 6 with GCC 6.1.0
Lots of error messages, not sure which ones are important:
```
$ grep error: spack-build.out
> collect2: error: ld returned 1 exit status
> /scratch/ajstewart/spack-stage/spack-stage-853rvo/qt-everywhere-opensource-src-5.9.1/qtbase/config.tests/unix/libudev/libudev.cpp:40:21: fatal error: libudev.h: No such file or directory
> collect2: error: ld returned 1 exit status
> collect2: error: ld returned 1 exit status
> /scratch/ajstewart/spack-stage/spack-stage-853rvo/qt-everywhere-opensource-src-5.9.1/qtbase/config.tests/unix/slog2/slog2.cpp:40:19: fatal error: slog2.h: No such file or directory
> /scratch/ajstewart/spack-stage/spack-stage-853rvo/qt-everywhere-opensource-src-5.9.1/qtbase/config.tests/unix/doubleconversion/doubleconversion.cpp:40:49: fatal error: double-conversion/double-conversion.h: No such file or directory
> /scratch/ajstewart/spack-stage/spack-stage-853rvo/qt-everywhere-opensource-src-5.9.1/qtbase/config.tests/unix/pcre2/pcre2.cpp:42:19: fatal error: pcre2.h: No such file or directory
> /scratch/ajstewart/spack-stage/spack-stage-853rvo/qt-everywhere-opensource-src-5.9.1/qtbase/config.tests/qpa/egl/egl.cpp:40:21: fatal error: EGL/egl.h: No such file or directory
> /scratch/ajstewart/spack-stage/spack-stage-853rvo/qt-everywhere-opensource-src-5.9.1/qtbase/config.tests/unix/openvg/openvg.cpp:40:23: fatal error: VG/openvg.h: No such file or directory
> /scratch/ajstewart/spack-stage/spack-stage-853rvo/qt-everywhere-opensource-src-5.9.1/qtbase/config.tests/unix/harfbuzz/harfbuzz.cpp:40:25: fatal error: harfbuzz/hb.h: No such file or directory
> /scratch/ajstewart/spack-stage/spack-stage-853rvo/qt-everywhere-opensource-src-5.9.1/qtbase/config.tests/unix/tslib/tslib.cpp:40:19: fatal error: tslib.h: No such file or directory
> /scratch/ajstewart/spack-stage/spack-stage-853rvo/qt-everywhere-opensource-src-5.9.1/qtbase/config.tests/unix/cups/cups.cpp:40:23: fatal error: cups/cups.h: No such file or directory
> /scratch/ajstewart/spack-stage/spack-stage-853rvo/qt-everywhere-opensource-src-5.9.1/qtbase/config.tests/unix/db2/db2.cpp:40:20: fatal error: sqlcli.h: No such file or directory
> /scratch/ajstewart/spack-stage/spack-stage-853rvo/qt-everywhere-opensource-src-5.9.1/qtbase/config.tests/unix/ibase/ibase.cpp:40:19: fatal error: ibase.h: No such file or directory
> /scratch/ajstewart/spack-stage/spack-stage-853rvo/qt-everywhere-opensource-src-5.9.1/qtbase/config.tests/unix/mysql/mysql.cpp:43:19: fatal error: mysql.h: No such file or directory
> /scratch/ajstewart/spack-stage/spack-stage-853rvo/qt-everywhere-opensource-src-5.9.1/qtbase/config.tests/unix/mysql/mysql.cpp:43:19: fatal error: mysql.h: No such file or directory
> /scratch/ajstewart/spack-stage/spack-stage-853rvo/qt-everywhere-opensource-src-5.9.1/qtbase/config.tests/unix/oci/oci.cpp:40:17: fatal error: oci.h: No such file or directory
> /scratch/ajstewart/spack-stage/spack-stage-853rvo/qt-everywhere-opensource-src-5.9.1/qtbase/config.tests/unix/odbc/odbc.cpp:43:17: fatal error: sql.h: No such file or directory
> /scratch/ajstewart/spack-stage/spack-stage-853rvo/qt-everywhere-opensource-src-5.9.1/qtbase/config.tests/unix/sqlite2/sqlite2.cpp:40:20: fatal error: sqlite.h: No such file or directory
> /scratch/ajstewart/spack-stage/spack-stage-853rvo/qt-everywhere-opensource-src-5.9.1/qtbase/config.tests/unix/tds/tds.cpp:40:22: fatal error: sybfront.h: No such file or directory
> /usr/include/linux/can.h:81:2: error: ‘sa_family_t’ does not name a type
> /scratch/ajstewart/spack-stage/spack-stage-853rvo/qt-everywhere-opensource-src-5.9.1/qtdeclarative/config.tests/d3d12/d3d12.cpp:40:19: fatal error: d3d12.h: No such file or directory
> /scratch/ajstewart/spack-stage/spack-stage-853rvo/qt-everywhere-opensource-src-5.9.1/qtgamepad/config.tests/sdl/main.cpp:37:17: fatal error: SDL.h: No such file or directory
> /scratch/ajstewart/spack-stage/spack-stage-853rvo/qt-everywhere-opensource-src-5.9.1/qt3d/config.tests/assimp/main.cpp:29:28: fatal error: assimp/cimport.h: No such file or directory
> /scratch/ajstewart/spack-stage/spack-stage-853rvo/qt-everywhere-opensource-src-5.9.1/qt3d/config.tests/fbx/main.cpp:29:20: fatal error: fbxsdk.h: No such file or directory
> /scratch/ajstewart/spack-stage/spack-stage-853rvo/qt-everywhere-opensource-src-5.9.1/qtwayland/config.tests/wayland/main.cpp:41:29: fatal error: wayland-version.h: No such file or directory
> /scratch/ajstewart/spack-stage/spack-stage-853rvo/qt-everywhere-opensource-src-5.9.1/qtwayland/config.tests/wayland_egl/main.cpp:41:25: fatal error: wayland-egl.h: No such file or directory
> /scratch/ajstewart/spack-stage/spack-stage-853rvo/qt-everywhere-opensource-src-5.9.1/qtwayland/config.tests/wayland/main.cpp:41:29: fatal error: wayland-version.h: No such file or directory
> /scratch/ajstewart/spack-stage/spack-stage-853rvo/qt-everywhere-opensource-src-5.9.1/qtconnectivity/config.tests/bluez/main.cpp:29:33: fatal error: bluetooth/bluetooth.h: No such file or directory
> /scratch/ajstewart/spack-stage/spack-stage-853rvo/qt-everywhere-opensource-src-5.9.1/qtlocation/config.tests/winrt/main.cpp:30:28: fatal error: windows.system.h: No such file or directory
```
Full [spack-build.out](https://github.com/LLNL/spack/files/1195531/spack-build.txt)

### LINKED ISSUE #4961: Unable to build qt on CentOS 7 with GCC 7.1.0
This one actually produced a useful error message:
```
WARNING: Could not find XKB config root, use -xkb-config-root to set a path to
XKB configuration data. This is required for keyboard input support.

ERROR: Feature 'opengl' was enabled, but the pre-condition 'features.opengl-desktop || features.opengl-dynamic || features.opengles2' failed.

ERROR: The OpenGL functionality tests failed!
```
Full [spack-build.out](https://github.com/LLNL/spack/files/1195614/spack-build.txt)

### PR DESCRIPTION
**Pull request by mjwoods**
Fixes #4960 
Fixes #4961

I tried installing different versions of qt on RHEL6 and RHEL7, and I found a number of issues that are fixed in this PR.

`spack install qt@3.3.8b` hung on my system, and it turned out that it was waiting for acceptance of a licence agreement (although the message only appeared in spack-build.out). The variant `+krellpatch` provides a patch that accepts the licence agreement automatically, and I moved this part of the patch into a separate patch that is applied with or without `+krellpatch` (e5de90d).

`spack install qt@3.3.8b` also failed later in the build due to a missing definition of type `ptrdiff_t`. The definition is provided by `stddef.h`, which I have included in the relevant files via a patch (8983b2b). The original code may have worked on older systems, but apparently not any more.

The default qt version (5.10.0) failed to install unless mesa libraries were available. Even though spack already sets configure option `-no-opengl` for variant `~opengl`, there is a qt component (webglplugin) that tries to link with opengl. To solve the problem, I disabled webglplugin for variant `~opengl` (72cdc2c).

Finally, `qt@3.3.8b` and `qt@4.8.6` both attempt to link libXext. (My test system has a minimal set of development packages, which helps to reveal hidden dependencies). To fix the problem, I added a new dependency to the qt package (5598e5c). I originally wanted to limit the dependency to `qt@3:4.99`, but I couldn't express this in combination with `when=sys.platform != 'darwin'`. For example, spack gives a syntax error for the expression `when='qt@3:4.99 platform!=darwin'`.

### CODE REVIEW

**Comment by davydden** [comment]
> I originally wanted to limit the dependency to qt@3:4.99, but I couldn't express this in combination with when=sys.platform != 'darwin'. For example, spack gives a syntax error for the expression when='qt@3:4.99 platform!=darwin

Just put it outside:

if sys.platform != ‘darwin’:
    Depends_on....

**Comment by mjwoods** [comment]
Good idea, @davydden . I have grouped a few X dependencies in an `if` block, so that they are ignored on OSX.

Is it safe to assume that qt should always be built without X support on OSX? Or should we provide a package variant `~X`?

**Comment by mjwoods** [comment]
I looked at some of your qt issues, @adamjstewart . I think we can close #4960 and #4961. I used compilers with newer minor versions from yours (because I had them ready to use), but I don't expect that slightly older compiler versions would change the results.

File: var/spack/repos/builtin/packages/qt/package.py [hunk 1]
```diff
@@ -123,10 +125,12 @@ class Qt(Package):
     # QtQml
     depends_on("python", when='@5.7.0:', type='build')
 
-    # OpenGL hardware acceleration
-    depends_on("mesa", when='@4:+opengl')
```
Comment (davydden) [inline var/spack/repos/builtin/packages/qt/package.py]: Sorry for confusing you, I think I misunderstood the original comment. I would keep it the way it is, including mesa. That is, just drop the last commit.

I think there is another PR which adds ‘gl’ virtual dependency that can be satisfied by Mesa or system provided libs.

Comment (davydden) [reply var/spack/repos/builtin/packages/qt/package.py]: Ps. You would keep as it’s indeed not possible to combine it with version constraint


if sys.platform != 'darwin':
     depends_on("libxext", when='@3:4.99')

Comment (mjwoods) [reply var/spack/repos/builtin/packages/qt/package.py]: Now I'm confused :confused: - do you want me to keep 05ce08de94 or revert it?

Comment (davydden) [reply var/spack/repos/builtin/packages/qt/package.py]: Sorry for confusion, I would keep part of it (new dependency), and revert the rest.

Comment (mjwoods) [reply var/spack/repos/builtin/packages/qt/package.py]: OK, please see c568d556a4d.

**Pull request merged**